# Copyright 2026 Qore Technologies, s.r.o.; Apache-2.0.
# Keep asynchronous batch ownership with the Core callback through cancellation.
# Release rejected batches and keep response conversion cleanup exception-safe.
--- a/src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pyx.pxi
@@ -32,7 +32,8 @@
 
 cdef class CallbackWrapper:
 
-    def __cinit__(self, object future, object loop, CallbackFailureHandler failure_handler):
+    def __cinit__(self, object future, object loop, CallbackFailureHandler failure_handler,
+                  _BatchOperationTag batch_operation_tag=None):
         self.context.functor.functor_run = self.functor_run
         self.context.waiter = <cpython.PyObject*>future
         self.context.loop = <cpython.PyObject*>loop
@@ -42,6 +43,7 @@
         # data path. We should make it as efficient as possible.
         self._reference_of_future = future
         self._reference_of_failure_handler = failure_handler
+        self._batch_operation_tag = batch_operation_tag
         # NOTE(lidiz) We need to ensure when Core invokes our callback, the
         # callback function itself is not deallocated. Otherwise, we will get
         # a segfault. We can view this as Core holding a ref.
@@ -53,12 +55,33 @@
             int success) noexcept:
         cdef CallbackContext *context = <CallbackContext *>functor
         cdef object waiter = <object>context.waiter
-        if not waiter.cancelled():
-            if success == 0:
-                (<CallbackFailureHandler>context.failure_handler).handle(waiter)
+        cdef CallbackWrapper wrapper = <CallbackWrapper>context.callback_wrapper
+        cdef grpc_event c_event
+        c_event.type = GRPC_OP_COMPLETE
+        c_event.success = success
+        c_event.tag = NULL
+        try:
+            try:
+                # Core, not the awaiting coroutine, determines buffer lifetime.
+                # Cancellation can finish the coroutine before this callback.
+                if wrapper._batch_operation_tag is not None:
+                    wrapper._batch_operation_tag.event(c_event)
+            except BaseException as error:
+                if not waiter.done():
+                    waiter.set_exception(error)
+                else:
+                    (<object>context.loop).call_exception_handler({
+                        "message": "gRPC batch cleanup failed after cancellation",
+                        "exception": error,
+                    })
             else:
-                waiter.set_result(None)
-        cpython.Py_DECREF(<object>context.callback_wrapper)
+                if not waiter.done():
+                    if success == 0:
+                        (<CallbackFailureHandler>context.failure_handler).handle(waiter)
+                    else:
+                        waiter.set_result(None)
+        finally:
+            cpython.Py_DECREF(<object>context.callback_wrapper)
 
     cdef grpc_completion_queue_functor *c_functor(self):
         return &self.context.functor
@@ -79,13 +102,20 @@
                                object loop):
     """The callback version of start batch operations."""
     cdef _BatchOperationTag batch_operation_tag = _BatchOperationTag(None, operations, None)
+    cdef object future = loop.create_future()
+    cdef CallbackFailureHandler failure_handler = CallbackFailureHandler(
+        'execute_batch', operations, ExecuteBatchError)
+    cdef CallbackWrapper wrapper
+    cdef grpc_event c_event
+    c_event.type = GRPC_OP_COMPLETE
+    c_event.success = 0
+    c_event.tag = NULL
     batch_operation_tag.prepare()
-
-    cdef object future = loop.create_future()
-    cdef CallbackWrapper wrapper = CallbackWrapper(
-        future,
-        loop,
-        CallbackFailureHandler('execute_batch', operations, ExecuteBatchError))
+    try:
+        wrapper = CallbackWrapper(future, loop, failure_handler, batch_operation_tag)
+    except BaseException:
+        batch_operation_tag.event(c_event)
+        raise
     cdef grpc_call_error error = grpc_call_start_batch(
         grpc_call_wrapper.call,
         batch_operation_tag.c_ops,
@@ -93,14 +123,15 @@
         wrapper.c_functor(), NULL)
 
     if error != GRPC_CALL_OK:
+        # Core rejected the batch and will never invoke its callback.
+        try:
+            batch_operation_tag.event(c_event)
+        finally:
+            cpython.Py_DECREF(wrapper)
         grpc_call_error_string = grpc_call_error_to_string(error).decode()
         raise ExecuteBatchError("Failed grpc_call_start_batch: {} with grpc_call_error value: '{}'".format(error, grpc_call_error_string))
 
     await future
-
-    cdef grpc_event c_event
-    # Tag.event must be called, otherwise messages won't be parsed from C
-    batch_operation_tag.event(c_event)
 
 
 cdef prepend_send_initial_metadata_op(tuple ops, tuple metadata):
--- a/src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pxd.pxi
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/aio/callback_common.pxd.pxi
@@ -44,6 +44,7 @@
     cdef CallbackContext context
     cdef object _reference_of_future
     cdef object _reference_of_failure_handler
+    cdef _BatchOperationTag _batch_operation_tag
 
     @staticmethod
     cdef void functor_run(
--- a/src/python/grpcio/grpc/_cython/_cygrpc/tag.pyx.pxi
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/tag.pyx.pxi
@@ -67,9 +67,20 @@
   cdef BatchOperationEvent event(self, grpc_event c_event):
     cdef Operation operation
     if 0 < self.c_nops:
-      for operation in self._operations:
-        operation.un_c()
-      gpr_free(self.c_ops)
+      first_error = None
+      try:
+        for operation in self._operations:
+          try:
+            operation.un_c()
+          except BaseException as error:
+            if first_error is None:
+              first_error = error
+      finally:
+        gpr_free(self.c_ops)
+        self.c_ops = NULL
+        self.c_nops = 0
+      if first_error is not None:
+        raise first_error
       return BatchOperationEvent(
           c_event.type, c_event.success, self._user_tag, self._operations)
     else:
--- a/src/python/grpcio/grpc/_cython/_cygrpc/operation.pyx.pxi
+++ b/src/python/grpcio/grpc/_cython/_cygrpc/operation.pyx.pxi
@@ -135,8 +135,10 @@
         &self._c_initial_metadata)
 
   cdef void un_c(self) except *:
-    self._initial_metadata = _metadata(&self._c_initial_metadata)
-    grpc_metadata_array_destroy(&self._c_initial_metadata)
+    try:
+      self._initial_metadata = _metadata(&self._c_initial_metadata)
+    finally:
+      grpc_metadata_array_destroy(&self._c_initial_metadata)
 
   def initial_metadata(self):
     return self._initial_metadata
@@ -163,20 +165,27 @@
     cdef size_t message_slice_length
     cdef void *message_slice_pointer
     if self._c_message_byte_buffer != NULL:
-      message_reader_status = grpc_byte_buffer_reader_init(
-          &message_reader, self._c_message_byte_buffer)
-      if message_reader_status:
-        message = bytearray()
-        while grpc_byte_buffer_reader_next(&message_reader, &message_slice):
-          message_slice_pointer = grpc_slice_start_ptr(message_slice)
-          message_slice_length = grpc_slice_length(message_slice)
-          message += (<char *>message_slice_pointer)[:message_slice_length]
-          grpc_slice_unref(message_slice)
-        grpc_byte_buffer_reader_destroy(&message_reader)
-        self._message = bytes(message)
-      else:
-        self._message = None
-      grpc_byte_buffer_destroy(self._c_message_byte_buffer)
+      try:
+        message_reader_status = grpc_byte_buffer_reader_init(
+            &message_reader, self._c_message_byte_buffer)
+        if message_reader_status:
+          try:
+            message = bytearray()
+            while grpc_byte_buffer_reader_next(&message_reader, &message_slice):
+              try:
+                message_slice_pointer = grpc_slice_start_ptr(message_slice)
+                message_slice_length = grpc_slice_length(message_slice)
+                message += (<char *>message_slice_pointer)[:message_slice_length]
+              finally:
+                grpc_slice_unref(message_slice)
+            self._message = bytes(message)
+          finally:
+            grpc_byte_buffer_reader_destroy(&message_reader)
+        else:
+          self._message = None
+      finally:
+        grpc_byte_buffer_destroy(self._c_message_byte_buffer)
+        self._c_message_byte_buffer = NULL
     else:
       self._message = None
 
@@ -206,16 +215,19 @@
         &self._c_error_string)
 
   cdef void un_c(self) except *:
-    self._trailing_metadata = _metadata(&self._c_trailing_metadata)
-    grpc_metadata_array_destroy(&self._c_trailing_metadata)
-    self._code = self._c_code
-    self._details = _decode(_slice_bytes(self._c_details))
-    grpc_slice_unref(self._c_details)
-    if self._c_error_string != NULL:
-      self._error_string = _decode(self._c_error_string)
+    try:
+      self._trailing_metadata = _metadata(&self._c_trailing_metadata)
+      self._code = self._c_code
+      self._details = _decode(_slice_bytes(self._c_details))
+      if self._c_error_string != NULL:
+        self._error_string = _decode(self._c_error_string)
+      else:
+        self._error_string = ""
+    finally:
+      grpc_metadata_array_destroy(&self._c_trailing_metadata)
+      grpc_slice_unref(self._c_details)
       gpr_free(<void*>self._c_error_string)
-    else:
-      self._error_string = ""
+      self._c_error_string = NULL
 
   def trailing_metadata(self):
     return self._trailing_metadata
